homeworks/nico778/5-remesh/src/geometry/halfedge.ts

```diff
@@ -0,0 +1,456 @@
+    
+    let eQ = new EdgeQueue((e1, e2) => {
+      return e1.error() < e2.error();
+    });
+
+    this.edges.forEach(e => {
+      if(e) {
+        eQ.push(e);
+      }
+    });
+    
+    while(reducedFaces > 0) {
+      let e = eQ.pop();
+      let bEdge = this.edges[e.idx];
+
+      if(e.removed) { 
+        continue; 
+      }
+
+      if(bEdge!.halfedge!.onBoundary) {
+        continue;
+      }
+
+      let v1 = e.halfedge!.vert;
+      let v2 = e.halfedge!.twin!.vert;
+
+      if(v1!.idx == v2!.idx) {
+        continue;
+      }
+
+      this.verts[v1!.idx] = null;
+      this.verts[v2!.idx] = null;
+
+      v1!.halfedges(h => {
+        h!.edge!.removed = true;
+        this.halfedges[h.idx] = null;
+        this.halfedges[h!.twin!.idx] = null;
+        this.edges[h!.edge!.idx] = null;
+        this.faces[h!.face!.idx] = null;
+      });
+
+      v2!.halfedges(h => {
+        h!.edge!.removed = true;
+        this.halfedges[h!.idx] = null;
+        this.halfedges[h!.twin!.idx] = null;
+        this.edges[h!.edge!.idx] = null;
+        this.faces[h!.face!.idx] = null;
+      });
+
+      //todo: new vert
+      let v = new Vertex(new Vector(0, 0, 0, 1));
+      
+      v.pos = e.bestVertex();
+      v.idx = this.verts.length;
+
+      this.verts.push(v);
+
+      //todo: test for bugs
+      let tE = this.testEdges(v1!, v2!);
+      let ecEdges: Edge[] = [];
+      let ecHalfedges: Halfedge[] = [];
+      
+      tE.forEach(h => {
+        let e = new Edge();
+        e.halfedge! = new Halfedge();
+        e.halfedge!.twin! = new Halfedge();
+        //still missing
+        e.idx = this.edges.length;
+        e.halfedge!.idx = this.halfedges.length;
+        e.halfedge!.edge = e;
+        e.halfedge!.twin!.idx = this.halfedges.length;
+        e.halfedge!.twin!.edge = e;
+        e.halfedge!.twin!.twin = e.halfedge!;
+
+        this.edges.push(e);
+        this.halfedges.push(e.halfedge);
+        this.halfedges.push(e.halfedge.twin);
+        ecEdges.push(e);
+        ecHalfedges.push(h!);
+      });
+
+      //reduce by 2 not 1
+      reducedFaces -= 2;
+
+      //todo: halfedge connectivity
+      this.updateHC(v, ecEdges, ecHalfedges);
+      eQ.push(...ecEdges);
+    }
+    this.resetIndices();
+  }
+
+testEdges(v1: Vertex, v2: Vertex) {
+  let h1, h2;
+  
+  if(v1!.halfedge!.twin!.vert == v2) {
+    h1 = v1!.halfedge!.twin!.next;
+  } else {
+    h1 = v1!.halfedge!;
+  }
+  if(h1!.next!.twin!.vert == v2) {
+    h2 = h1!.next!.twin!.next;
+  } else {
+    h2 = h1!.next;
+  }
+
+  let tE = [h2];
+  let i = h2;
+  //still something going wrong here
+  while(true) {
+    //bug causes crash
+    //fix: assigned wrong, use twin
+    if(i!.next!.twin!.next!.twin!.vert == v1 || i!.next!.twin!.next!.twin!.vert == v2) {
+      i = i!.next!.twin!.next!.twin!.next;
+    } else {
+      i = i!.next!.twin!.next;
+    }
+    if(i == h2) {
+      break;
+    }
+    tE.push(i);
+  }
```

Comment:
I guess this is where a bug could easily be identified as incorrect implementation result in an infinite loop :)